ocaml/xapi/cert_distrib.ml

```diff
@@ -0,0 +1,247 @@
+          CollectResult cert
+      | Write (strategy, certs) ->
+          write_certs_fs strategy certs ;
+          WriteResult
+      | GenBundle ->
+          regen_bundle () ; GenBundleResult
+    in
+    string_of_result r
+
+  let result_or_fail x =
+    result_of_string x |> function
+    | Some x ->
+        x
+    | None ->
+        raise_internal
+          ~details:(Printf.sprintf "result string: %s" x)
+          "result_or_fail: failed to parse result"
+
+  let remote_call_sync host payload rpc session_id =
+    XenAPI.Host.cert_distrib_atom rpc session_id host
+      (string_of_payload payload)
+    |> result_or_fail
+
+  let unexpected_result name r =
+    raise_internal
+      ~details:(Printf.sprintf "r = %s" (dbg_of_result r))
+      (Printf.sprintf "%s: unexpected result" name)
+
+  let remote_collect_cert host rpc session_id =
+    remote_call_sync host Collect rpc session_id |> function
+    | CollectResult cert ->
+        cert
+    | r ->
+        unexpected_result "remote_collect_cert" r
+
+  let remote_write_certs_fs strategy certs host rpc session_id =
+    remote_call_sync host (Write (strategy, certs)) rpc session_id |> function
+    | WriteResult ->
+        ()
+    | r ->
+        unexpected_result "remove_write_certs_fs" r
+
+  let remote_regen_bundle host rpc session_id =
+    remote_call_sync host GenBundle rpc session_id |> function
+    | GenBundleResult ->
+        ()
+    | r ->
+        unexpected_result "remote_regen_bundle" r
+end
+
+let local_exec = Worker.local_exec
+
+let go ~__context ~from_hosts ~to_hosts ~existing_cert_strategy =
+  (* here we coordinate the certificate distribution. from a high level
+   * we do the following:
+   * a) collect certs from [from_hosts], aggregating them on the master.
+   *    the cert collected will be [Worker.my_cert]
+   * b) tell [to_hosts] to write the certs collected in (a) to [Worker.pool_certs]).
+   *    the original contents of [Worker.pool_certs] will either be removed or
+   *    simply added to, depending on [existing_cert_strategy]
+   * c) tell [to_hosts] to regenerate their trust bundles. the old bundle is
+   *    removed completely. see 'update-ca-bundle.sh'. only at this point would we
+   *    expect 'things to go wrong', e.g. new connections fail, if the certs have not
+   *    been set up correctly
+   * NB: if any individual call fails, then we don't continue with the distribution.
+   *     for example: suppose that the master is unable to obtain the cert from a host,
+   *     then we are guaranteed not to modify the trusted certs on any hosts. however
+   *     we do not guarantee 'atomicity', so if regenerating the bundle on one host
+   *     fails, then state across the pool will most likely become inconsistent, and
+   *     manual intervention may be required *)
+  let uuid host = Db.Host.get_uuid ~__context ~self:host in
+
+  Helpers.call_api_functions ~__context @@ fun rpc session_id ->
+  let blobs =
```

Comment:
I'm thinking about how to change this so we can get rollback when cycling the internal certificates.

We could change this function here to request new certificates to hosts (and save them in a temp location).

After that, change the distribution so it's a two step process: 1. backup the certs and copy the new certificates (arm change) 2. remove the backup (finish_distribution) or remove the newly-copied certs and restore from backup (rollback)